spec/google/gax/bundling_spec.rb

```diff
@@ -0,0 +1,661 @@
+          expect(event.set?).to eq(true)
+          expect(event.result).to eq(
+            mismatched_result
+          )
+          previous_event = event
+        end
+      end
+    end
+
+    context 'method `schedule`' do
+      it 'passes kwargs correctly' do
+        an_elt = 'dummy_msg'
+        options = Google::Gax::BundleOptions.new(element_count_threshold: 1)
+        bundle_id = 'an_id'
+        bundler = Google::Gax::Executor.new(options)
+        event = bundler.schedule(
+          return_kwargs,
+          bundle_id,
+          SIMPLE_DESCRIPTOR,
+          bundled_builder([an_elt]),
+          kwargs: { an_option: 'a_value' }
+        )
+        expect(event.result[:an_option]).to eq('a_value')
+      end
+    end
+
+    context 'bundles are triggered to run correctly' do
+      it 'api call not invoked until element threshold' do
+        an_elt = 'dummy_msg'
+        an_id = 'bundle_id'
+        api_call = return_request
+        threshold = 3
+        options =
+          Google::Gax::BundleOptions.new(element_count_threshold: threshold)
+        bundler = Google::Gax::Executor.new(options)
+        1.upto(3) do |i|
+          event = bundler.schedule(
+            api_call,
+            an_id,
+            SIMPLE_DESCRIPTOR,
+            bundled_builder([an_elt])
+          )
+
+          expect(event.canceller).to_not be_nil
+          if i < threshold
+            expect(event.set?).to eq(false)
+            expect(event.result).to be_nil
+          else
+            expect(event.set?).to eq(true)
+            expect(event.result).to eq(
+              bundled_builder([an_elt] * threshold)
+            )
+          end
+        end
+      end
+
+      it 'api call not invoked until byte threshold' do
+        an_elt = 'dummy_msg'
+        an_id = 'bundle_id'
+        api_call = return_request
+        elts_for_threshold = 3
+        threshold = elts_for_threshold * an_elt.bytesize
+        options =
+          Google::Gax::BundleOptions.new(request_byte_threshold: threshold)
+        bundler = Google::Gax::Executor.new(options)
+        1.upto(elts_for_threshold) do |i|
+          event = bundler.schedule(
+            api_call,
+            an_id,
+            SIMPLE_DESCRIPTOR,
+            bundled_builder([an_elt])
+          )
+
+          expect(event.canceller).to_not be_nil
+          if i < elts_for_threshold
+            expect(event.set?).to eq(false)
+            expect(event.result).to be_nil
+          else
+            expect(event.set?).to eq(true)
+            expect(event.result).to eq(
+              bundled_builder([an_elt] * elts_for_threshold)
+            )
+          end
+        end
+      end
+
+      it 'api call not invoked until time threshold' do
+        asleep = true
+
+        # This needs to be any instance because the instance of Kernal inside
+        # the thread will not be the main thread's instance of kernal
+        allow_any_instance_of(Kernel).to receive(:sleep) do |_|
```

Comment:
Using 'allow_any_instance_of' on the Kernel is dangerous and can cause unpredictable behavior as seen in this builds travis test failing. I am going to make the test less robust until I find a better way to mock time accross threads.